services/src/main/java/org/keycloak/protocol/oidc/endpoints/LogoutEndpoint.java

```diff
@@ -229,8 +258,22 @@ public Response logout(@QueryParam(OIDCLoginProtocol.REDIRECT_URI_PARAM) String
         LoginFormsProvider loginForm = session.getProvider(LoginFormsProvider.class)
                 .setAuthenticationSession(logoutSession);
 
-        // Client was not sent in id_token_hint or has consentRequired. Logout confirmation screen will be displayed to the user in this case
-        if (client == null || client.isConsentRequired()) {
+        // Check if we have session in the browser. If yes and it is different session than referenced by id_token_hint, the confirmation should be displayed
+        AuthenticationManager.AuthResult authResult = AuthenticationManager.authenticateIdentityCookie(session, realm, false);
+        if (authResult != null) {
+            if (idToken != null && idToken.getSessionState() != null && !idToken.getSessionState().equals(authResult.getSession().getId())) {
+                forcedConfirmation = true;
+            }
+        } else {
+            // Skip confirmation in case that valid redirect URI was setup for given client_id and there is no session in the browser as well as no id_token_hint.
+            // We can do automatic redirect as there is no logout needed at all for this scenario (Session was probably already logged-out before)
+            if (encodedIdToken == null && client != null && validatedRedirectUri != null) {
+                confirmationNeeded = false;
+            }
+        }
+
+        // Logout confirmation screen will be displayed to the user in this case
+        if (confirmationNeeded || forcedConfirmation) {
```

Comment:
Out of curiosity what is the difference between `confirmationNeeded` and `forcedConfirmation`?

Reply:
Those are just 2 helper boolean variables. First is true by default and switches to false once the conditions are evaluated that confirmation not needed (EG. valid id_token). Second is true by default and can switch to false if confirmation should be always shown regardless of any other things (Like the case with unknown client discussed above).

Was just easier for me to have 2 boolean variables to be on the safe side and not miss any obvious combination when the confirmation should be shown and is not (also recall some recent example from security training related to this :-)